File: extract.js
const { execSync } = require("child_process");
const program = require("commander");
const fs = require("fs");
const path = require("path");

program
  .usage("[options] <file ...>")
  .option("-s, --source <path>", "源目录")
  .option("-d, --destination <path>", "目标目录");

program.parse(process.argv);

const { source, destination } = program;

fs.readdir(source, "utf-8", function callback(err, files) {
  if (err) {
    return console.log(err);
  }

  const len = files.length;

  for (let i = 0; i < len; i++) {
    const file = files[i];
    const absPath = source + file;
    const extname = path.extname(absPath);
    if (extname !== ".rar" && extname !== ".zip") {
      continue;
    }
    // 因为文件名包含三位数目录和其他汉字，所以获取三位数目录, 作为解压后目录
    const rename = file.match(/\d{3}/g)[0];
    const filename = rename + extname;
    const destFile = `${destination + filename}`;
    try {
      let extractCmd;
      const destFolder = `${destination + rename}`;
      if (extname === ".rar") {
        extractCmd = `unrar e ${destFile} ${destFolder}`;
      } else {
        extractCmd = `ditto -x -k ${destFile} ${destFolder}`;
      }

      fs.copyFileSync(absPath, destFile);
      execSync(`mkdir ${destFolder}`);
      execSync(extractCmd);
      // const files = fs.readdirSync(`${destFolder}`, "utf-8");
      // if (files.length === 1) {
      //   execSync(`mv -f ${destFolder + "/" + files[0].replace("\\", "\\\\") + "/*"} ${destFolder}`);
      //   execSync(`rm -rf ${destFolder + "/" + files[0].replace("\\", "\\\\")}`);
      // }
    } catch (error) {
      console.log(filename + ": " + error.message);
    } finally {
      execSync(`rm ${destFile}`);
    }
  }
});

// unzip src -d dest
// unrar e src dest

File: extract.go
package main

import (
	"flag"
	"fmt"
	"io/ioutil"
	"os"
	"os/exec"
	"path"
	"path/filepath"
	"regexp"
	"time"
)

func createDirIfNotExist(dir string) {
	if _, err := os.Stat(dir); os.IsNotExist(err) {
		err = os.MkdirAll(dir, 0755)
		if err != nil {
			panic(err)
		}
	}
}

func main() {

	start := time.Now()

	var src string
	var dest string
	flag.StringVar(&src, "s", "src", "源目录")
	flag.StringVar(&dest, "d", "dest", "目标目录")
	flag.Parse()

	files, err := ioutil.ReadDir(src)
	if err != nil {
		panic(err)
	}
	for _, file := range files {
		ext := filepath.Ext(file.Name())

		if ext != ".rar" && ext != ".zip" {
			continue
		}
		re, _ := regexp.Compile("[0-9]{3}")
		bytes := re.Find([]byte(file.Name()))
		rename := string(bytes)
		folder := path.Join(dest, rename)
		createDirIfNotExist(folder)
		var cmd *exec.Cmd
		if ext == ".rar" {
			cmd = exec.Command("unrar", "e", path.Join(src, file.Name()), folder)
		} else {
			cmd = exec.Command("ditto", "-x", "-k", path.Join(src, file.Name()), folder)
		}
		err := cmd.Run()
		if err != nil {
			fmt.Printf("%v\n", err)
		}
		childFiles, err1 := ioutil.ReadDir(folder)
		if err1 != nil {
			panic(err1)
		}

		for _, childFile := range childFiles {
			if childFile.IsDir() {
				err = filepath.Walk(path.Join(folder, childFile.Name()), func(p string, info os.FileInfo, err error) error {
					extName := filepath.Ext(p)
					newName := path.Join(folder, info.Name())
					if extName == ".jpg" || extName == ".jpeg" || extName == ".png" {
						os.Rename(p, newName)
					}
					return nil
				})
				os.RemoveAll(path.Join(folder, childFile.Name()))
			}

		}
	}
	t := time.Now()
	elapsed := t.Sub(start)
	println(elapsed.Milliseconds())
}
